File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.ts
function lengthOfLongestSubstring(s: string): number {
  let start = 0;
  let maxLength = 0;
  const charMap = {};
  for (let end = 0; end < s.length; end++) {
    let lastChar = s[end];
    if (lastChar in charMap) {
      start = Math.max(start, charMap[lastChar] + 1)
    }
    charMap[lastChar] = end;
    maxLength = Math.max(maxLength, end - start + 1);
  }
  return maxLength;
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start, end, max_length = 0, 0, 0
        char_map = {}
        while end < len(s):
            last_char = s[end]
            if last_char in char_map:
                start = max(start, char_map[last_char] + 1)
            char_map[last_char] = end
            max_length = max(max_length, end - start + 1)
            end += 1
        return max_length